dhall/src/Dhall/Import.hs

```diff
@@ -1019,62 +1000,105 @@ getCacheBaseDirectory = alternative₀ <|> alternative₁ <|> alternative₂
 
 -- If the URL contains headers typecheck them and replace them with their normal
 -- forms.
-normalizeHeaders :: URL -> StateT Status IO URL
-normalizeHeaders url@URL { headers = Just headersExpression } = do
+normalizeHeadersIn :: URL -> StateT Status IO URL
+normalizeHeadersIn url@URL { headers = Just headersExpression } = do
     Status { _stack } <- State.get
     loadedExpr <- loadWith headersExpression
+    let handler (e :: SomeException) = throwMissingImport (Imported _stack e)
+    normalized <- liftIO $ handle handler (normalizeHeaders loadedExpr)
+    return url { headers = Just (fmap absurd normalized) }
+
+normalizeHeadersIn url = return url
+
+-- | Empty origin headers used for remote contexts
+--   (and fallback when nothing is set in env or config file)
+emptyOriginHeaders :: Expr Src Import
+emptyOriginHeaders = ListLit (Just (fmap absurd originHeadersTypeExpr)) mempty
+
+-- | A fake Src to annotate headers expressions with.
+--   We need to wrap headers expressions in a Note for nice error reporting,
+--   and because `?` handling only catches SourcedExceptions
+headersSrc :: Src
+headersSrc = Src {
+        srcStart = SourcePos {
+            sourceName = fakeSrcName,
+            sourceLine = mkPos 1,
+            sourceColumn = mkPos 1
+        },
+        srcEnd = SourcePos {
+            sourceName = fakeSrcName,
+            sourceLine = mkPos 1,
+            sourceColumn = mkPos (Text.length fakeSrcText)
+        },
+        srcText = fakeSrcText
+    }
+  where
+    fakeSrcText = "«Origin Header Configuration»"
+    fakeSrcName = "[builtin]"
+
+-- | Load headers only from the environment (used in tests)
+envOriginHeaders :: Expr Src Import
+envOriginHeaders = Note headersSrc (Embed (Import (ImportHashed Nothing (Env "DHALL_HEADERS")) Code))
+
+-- | Load headers in env, falling back to config file
+defaultOriginHeaders :: IO (Expr Src Import)
+defaultOriginHeaders = do
+    fromFile <- originHeadersFileExpr
+    return (Note headersSrc (ImportAlt envOriginHeaders (Note headersSrc fromFile)))
```

Comment:
I think we won't be able to use `ImportAlt` here since, as you noted, it won't let us remember which import succeeded.  i think you'd need to change the type from `IO (Expr Src Import)` to `IO [Import]` to get this to work.  It would also improve the error messages, too, I think

Reply:
I went down this path, but I couldn't see how to make it work. The fallback code is implemented by `loadWith`. I considered extracting the logic from the `ImportAlt` branch, as a function with signature `StateT Status IO (Expr Src Void) -> StateT Status IO (Expr Src Void) -> StateT Status IO (Expr Src Void)`. I couldn't just take two `Expr Src Void`, since we need to provide a different _stack in each branch (based on the header we're importing).

But it seems we need to use the full `loadWith`, since that's also the code which does the `Cycle` detection. In particular, if we manually construct `_stack = headerImport :| NonEmpty.tail (_stack status)` then that would mean we'd need to also implement the cycle detection here too. That seems messy and repetitive.

Am I missing something?

I did try an alternate, which is to just literally use the parent stack. It removed the need to special-case reentrant loads, which is nice. It requires a runtime assertion (because we can't prove the parent stack is nonempty with types). It also shouldn't ever hit a referentially opaque import error, although that's an implementation detail (because we cache on the first load). I've made that change in c77f5a7efbdd7995fa5a5a2f1816a928f18690b7

Reply:
@timbertson: Oh, I just figured out a (hacky) way to do this, but at least it doesn't require the partial function: instead of trying to figure out which `headers.dhall` we imported, just add both of them to the stack

Reply:
Hmm, would errors then report that e.g. ~/.config/dhall/headers.dhall was imported from `env:DHALL_HEADERS` (or vice versa)? The partial function shouldn't ever fail (despite being a bit yuck), right? 🤞

Reply:
Yeah, you're right.  Then we'll stick with what you have for now